insights/tests/client/connection/test_test_connection.py

```diff
@@ -0,0 +1,175 @@
+    with caplog.at_level(ERROR, logger="insights.client.connection"):
+        insights_connection.test_connection()
+
+    assert not caplog.record_tuples
+
+
+@parametrize_exceptions
+@patch("insights.client.connection.InsightsConnection._test_urls")
+def test_test_connection_error_log_fail(
+    test_urls, exception, insights_connection, caplog
+):
+    """An error message is logged on the ERROR level."""
+    test_urls.side_effect = exception
+
+    with caplog.at_level(ERROR, logger="insights.client.connection"):
+        insights_connection.test_connection()
+
+    assert caplog.record_tuples == [
+        (
+            "insights.client.connection",
+            ERROR,
+            "Connectivity test failed! Please check your network configuration",
+        )
+    ]
+
+
+@patch("insights.client.connection.InsightsConnection._test_urls")
+def test_test_connection_print_no_catch(test_urls, insights_connection, capsys):
+    """The connection test doesn’t print anything in case of an unknown exception."""
+    test_urls.side_effect = Exception
+
+    try:
+        insights_connection.test_connection()
+    except Exception:
+        pass
+
+    out, err = capsys.readouterr()
+    assert not out
+    assert not err
+
+
+@patch("insights.client.connection.InsightsConnection._test_urls")
+def test_test_connection_print_success(test_urls, insights_connection, capsys):
+    """The connection test prints a message pointing to a log file if all API calls succeed."""
+    insights_connection.test_connection()
+
+    out, err = capsys.readouterr()
+    assert out == "See %s for more details.\n" % (LOGGING_FILE,)
+    assert not err
+
+
+@parametrize_exceptions
+@patch("insights.client.connection.InsightsConnection._test_urls")
+def test_test_connection_print_fail(test_urls, exception, insights_connection, capsys):
+    """The connection test prints a message pointing to a log file if an API call fails."""
+    message = "request exception"
+    test_urls.side_effect = exception(message)
+
+    insights_connection.test_connection()
+    out, err = capsys.readouterr()
+    assert out == "%s\nAdditional information may be in %s\n" % (
+        message,
+        LOGGING_FILE,
+    )
+    assert not err
+
+
+@patch("insights.client.connection.InsightsConnection._test_urls")
+def test_test_connection_no_catch(test_urls, insights_connection):
+    """The connection test doesn’t catch and recreate an unknown exception."""
+    expected = Exception()
+    test_urls.side_effect = expected
+
+    with raises(Exception) as caught:
+        insights_connection.test_connection()
+
+    assert caught.value is expected
+
+
+@patch("insights.client.connection.InsightsConnection._test_urls")
+def test_test_connection_return_code_success(test_urls, insights_connection):
+    """The connection test returns a zero code if all API calls succeed."""
+    result = insights_connection.test_connection()
+    assert result == 0
```

Comment:
Could this be merged with `test_test_connection_test_urls_success`?

Reply:
I am not sure this is a correct thing to do. The `test_test_connection_test_urls_*` tests only verify the calls to the `_test_urls` method. Hence the `test_urls` in their names. The `test_test_connection_return_code_*` ones verify the return value, which is a very different thing.

We can remove the two `return_code` tests from this pull request as they are not related to the error messages. I put it there however, to cover the case when the function doesn’t throw (or let pass uncaught) any exception.

Reply:
More thoughts:

To convey the idea that the return value check is there only for the tests to have at least some meaningful `assert`, I put the `result > 0` rather than `result == 1` in the other return value test. They are there only to verify the call doesn’t fail with an error, not testing the return values precisely.

Maybe `assert True` would convey this even better. But that is implicitly tested by all other tests calling the `_test_urls` method. Seeing that I think the return value tests can be safely removed.